### PR TITLE
ISSUE-678 JMAP - EmailRecoveryAction/set create

### PR DESCRIPTION
**Pull request by vttranlina**
resolve https://github.com/linagora/tmail-backend/issues/678

### CODE REVIEW

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/EmailRecoveryActionSetMethodContract.scala [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+           |                        ]
+           |                    }
+           |                }
+           |            },
+           |            "c1"
+           |        ]
+           |    ]
+           |}""".stripMargin)
+      .when
+      .post.prettyPeek()
```
Comment (Arsnael) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/EmailRecoveryActionSetMethodContract.scala:63]: debug spotted, dont forget it later :)

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CustomMethod.scala [hunk 2]
```diff
@@ -23,6 +23,7 @@ object CapabilityIdentifier {
   val LINAGORA_FIREBASE: CapabilityIdentifier = "com:linagora:params:jmap:firebase:push"
   val LINAGORA_CALENDAR: CapabilityIdentifier = "com:linagora:params:calendar:event"
   val LINAGORA_TEAM_MAILBOXES: CapabilityIdentifier = "com:linagora:params:jmap:team:mailboxes"
+  val LINAGORA_MESSAGE_VAULT: CapabilityIdentifier = "com:linagora:params:messages:vault"
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/CustomMethod.scala:26]: How about `com:linagora:params:jmap:messages:vault`

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/EmailRecoveryActionSetMethod.scala [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+  private def parseCreateRequest(json: JsObject): Either[Exception, EmailRecoveryActionCreationRequest] =
+    for {
+      validJsObject <- EmailRecoveryActionCreation.validateProperties(json)
+      parsedRequest <- EmailRecoveryActionSerializer.deserializeSetCreationRequest(validJsObject).asEither
+        .left.map(errors => new IllegalArgumentException(ResponseSerializer.serialize(JsError(errors)).toString))
+    } yield parsedRequest
+
+  private def submitTask(clientId: EmailRecoveryActionCreationId, userToRestore: Username, creationRequest: EmailRecoveryActionCreationRequest): SMono[CreationResult] = {
+    val taskId: TaskId = taskManager.submit(new DeletedMessagesVaultRestoreTask(restoreService, userToRestore, creationRequest.asQuery()))
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/EmailRecoveryActionSetMethod.scala:116]: This blocking call should be wrapped...

**Comment by vttranlina** [comment]
TODO

- [ ] Adapt James make public some methods & RestoreService should support query `limit`

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/EmailRecoveryActionSetMethodContract.scala [hunk 4]
```diff
@@ -24,44 +126,163 @@ trait EmailRecoveryActionSetMethodContract {
...
+      .jsonPath()
+      .get("methodResponses[0][1].created.clientId1.id").toString
+
+    `given`()
+      .spec(webAdminApi)
+      .get(taskId)
+      
+    .`then`()
+      .statusCode(SC_OK)
+      .body("taskId", Matchers.is(taskId))
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/EmailRecoveryActionSetMethodContract.scala:242]:
```suggestion
    `given`()
      .spec(webAdminApi)
      .get(taskId)
    .`then`()
      .statusCode(SC_OK)
      .body("taskId", Matchers.is(taskId))
```

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/EmailRecoveryActionSetMethodContract.scala [hunk 5]
```diff
@@ -73,15 +294,1246 @@ trait EmailRecoveryActionSetMethodContract {
...
+      .deliveryDate(deliveryDate)
+      .deletionDate(deletionDate)
+      .sender(sender)
+      .recipients(recipients.toList.asJava)
+      .hasAttachment(hasAttachment)
+      .size(contentLength)
+      .subject(subject)
+      .build
+
+  def awaitToRestoreTaskCompleted(taskId: String): ValidatableResponse =
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/EmailRecoveryActionSetMethodContract.scala:1523]:
```suggestion
  def awaitRestoreTaskCompleted(taskId: String): ValidatableResponse =
```

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/EmailRecoveryActionSetMethodContract.scala [hunk 6]
```diff
@@ -73,15 +294,1246 @@ trait EmailRecoveryActionSetMethodContract {
...
+      .body("status", Matchers.is("completed"))
+
+  def listAllMessageResult(guiceJamesServer: GuiceJamesServer, username: Username): java.util.List[MessageId] =
+    guiceJamesServer.getProbe(classOf[MailboxProbeImpl])
+      .searchMessage(MultimailboxesSearchQuery.from(SearchQuery.of(SearchQuery.all())).build, username.asString(), 100)
+      .asScala
+      .toList.asJava
+
+  def assertListAllMessageHasSize(guiceJamesServer: GuiceJamesServer, username: Username, size: Int): Unit =
+    await().atMost(30, TimeUnit.SECONDS).until(() => listAllMessageResult(guiceJamesServer, username).size() == size)
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/EmailRecoveryActionSetMethodContract.scala:1537]:
```suggestion
    await().atMost(30, TimeUnit.SECONDS).untilAsserted(() => assertThat(listAllMessageResult(guiceJamesServer, username)).hasSize(size))
```

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/EmailRecoveryActionSetMethodContract.scala [hunk 7]
```diff
@@ -73,15 +294,1246 @@ trait EmailRecoveryActionSetMethodContract {
...
+      .spec(webAdminApi)
+      .get(taskId + "/await")
+    .`then`()
+      .body("status", Matchers.is("completed"))
+
+  def listAllMessageResult(guiceJamesServer: GuiceJamesServer, username: Username): java.util.List[MessageId] =
+    guiceJamesServer.getProbe(classOf[MailboxProbeImpl])
+      .searchMessage(MultimailboxesSearchQuery.from(SearchQuery.of(SearchQuery.all())).build, username.asString(), 100)
+      .asScala
+      .toList.asJava
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/EmailRecoveryActionSetMethodContract.scala:1534]:
```suggestion
  def listAllMessageResult(guiceJamesServer: GuiceJamesServer, username: Username): java.util.Collection[MessageId] =
    guiceJamesServer.getProbe(classOf[MailboxProbeImpl])
      .searchMessage(MultimailboxesSearchQuery.from(SearchQuery.of(SearchQuery.all())).build, username.asString(), 100)
```

**Comment by chibenwa** [comment]
> Adapt James make public some methods & RestoreService should support query limit

Separate PR please

**Review by Arsnael — COMMENTED** [review]
Read it, nothing to add, good job :)

**Comment by vttranlina** [comment]
> Adapt James make public some methods & RestoreService should support query limit

https://github.com/apache/james-project/pull/1572

**Comment by vttranlina** [comment]
> ```
> [ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:4.6.1:compile (scala-compile-first) on project jmap-extensions: Execution scala-compile-first of goal net.alchim31.maven:scala-maven-plugin:4.6.1:compile failed: Compilation failed: InterfaceCompileFailed -> [Help 1]
> ```

we wait until the commit was merged first https://github.com/apache/james-project/pull/1572/commits/4efb8f20c80cb8823bc0ce4705607e63508ece3b , then update james to tmail, and rebase this pr

**Review by quantranhong1999 — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by vttranlina** [comment]
rebase master & squash

**Pull request merged**